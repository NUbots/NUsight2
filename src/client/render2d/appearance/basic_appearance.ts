import { observable } from 'mobx'

type Fill = { color: string, alpha: number }
type Stroke = { color: string, alpha: number, width: number }

export type BasicAppearanceOpts = {
  fill?: Partial<Fill>,
  stroke?: Partial<Stroke>
}

export class BasicAppearance {
  @observable fill?: Fill
  @observable stroke?: Stroke

  constructor({ fill, stroke }: { fill?: Fill, stroke?: Stroke }) {
    this.fill = fill
    this.stroke = stroke
  }

  static of({ fill, stroke }: BasicAppearanceOpts = {}): BasicAppearance {
    return new BasicAppearance({
      fill: fill && { color: '#000000', alpha: 1, ...fill },
      stroke: stroke && { color: '#000000', alpha: 1, width: 1, ...stroke },
    })
  }
}
