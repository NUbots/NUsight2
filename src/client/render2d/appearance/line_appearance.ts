import { observable } from 'mobx'


export type LineAppearanceOpts = {
  stroke?: {
    color: string,
    alpha?: number,
    width?: number,
    cap?: 'butt' | 'round' | 'square'
    dashOffset?: number
    join?: 'bevel' | 'round' | 'miter'
  }
}

export class LineAppearance {
  @observable stroke: {
    color: string,
    alpha: number,
    width: number,
    cap: 'butt' | 'round' | 'square'
    dashOffset: number
    join: 'bevel' | 'round' | 'miter'
  } = {
    color: '#000000',
    alpha: 1,
    width: 1,
    cap: 'butt',
    dashOffset: 0,
    join: 'miter',
  }

  constructor(opts: LineAppearanceOpts) {
    Object.assign(this.stroke, opts.stroke)
  }

  static of(opts: LineAppearanceOpts = {}): LineAppearance {
    return new LineAppearance(opts)
  }
}
