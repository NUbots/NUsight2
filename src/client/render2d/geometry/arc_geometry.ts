import { observable } from 'mobx'

import { Vector2 } from '../../math/vector2'

export class ArcGeometry {
  @observable origin: Vector2
  @observable radius: number
  @observable startAngle: number
  @observable endAngle: number
  @observable anticlockwise: boolean

  constructor(opts: ArcGeometry) {
    this.origin = opts.origin
    this.radius = opts.radius
    this.startAngle = opts.startAngle
    this.endAngle = opts.endAngle
    this.anticlockwise = opts.anticlockwise
  }

  static of({
              origin = Vector2.of(0, 0),
              radius = 1,
              startAngle = 0,
              endAngle = 2 * Math.PI,
              anticlockwise = false,
            }: Partial<ArcGeometry> = {}): ArcGeometry {
    return new ArcGeometry({
      origin,
      radius,
      startAngle,
      endAngle,
      anticlockwise,
    })
  }
}
