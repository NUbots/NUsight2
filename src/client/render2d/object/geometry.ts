import { ArcGeometry } from '../geometry/arc_geometry'
import { ArrowGeometry } from '../geometry/arrow_geometry'
import { CircleGeometry } from '../geometry/circle_geometry'
import { LineGeometry } from '../geometry/line_geometry'
import { PathGeometry } from '../geometry/path_geometry'
import { PolygonGeometry } from '../geometry/polygon_geometry'
import { TextGeometry } from '../geometry/text_geometry'

export type Geometry =
  ArcGeometry
  | ArrowGeometry
  | CircleGeometry
  | LineGeometry
  | PathGeometry
  | PolygonGeometry
  | TextGeometry
