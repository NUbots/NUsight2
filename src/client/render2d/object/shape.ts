import { observable } from 'mobx'

import { Appearance } from '../appearance/appearance'
import { BasicAppearance } from '../appearance/basic_appearance'

import { Geometry } from './geometry'

export type ShapeOpts<T extends Geometry> = {
  appearance: Appearance
  geometry: T
}

export class Shape<T extends Geometry> {
  @observable appearance: Appearance
  @observable geometry: T

  constructor(opts: ShapeOpts<T>) {
    this.appearance = opts.appearance
    this.geometry = opts.geometry
  }

  static of<T extends Geometry>(geometry: T, appearance: Appearance = BasicAppearance.of()) {
    return new Shape<T>({
      appearance,
      geometry,
    })
  }
}
