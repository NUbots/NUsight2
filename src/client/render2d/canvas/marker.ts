import { MarkerGeometry } from '../geometry/marker_geometry'
import { Shape } from '../object/shape'

import { applyAppearance } from './util'

export function renderMarker(ctx: CanvasRenderingContext2D, shape: Shape<MarkerGeometry>): void {

  const { x, y, radius, heading } = shape.geometry

  const headingAngle = Math.atan2(heading.y, heading.x)
  const arcDistance = 3 * Math.PI * 0.5

  // By default, the arc startAngle begins on the positive x-axis and rotates clockwise. If the startAngle and
  // endAngle are offset by a quadrant, the arc will point toward the positive x-axis instead of starting there.
  const startAngleOffset = Math.PI * 0.25
  const startAngle = headingAngle + startAngleOffset
  const endAngle = headingAngle + arcDistance + startAngleOffset

  ctx.beginPath()
  ctx.arc(
    x,
    y,
    radius,
    startAngle,
    endAngle,
  )

  // The diagonal length of a unit square.
  const sqrt2 = Math.sqrt(2)

  // Convert the heading to absolute canvas coordinates.
  ctx.lineTo(
    x + sqrt2 * radius * heading.x,
    y + sqrt2 * radius * heading.y,
  )
  ctx.closePath()

  applyAppearance(ctx, shape.appearance)

  ctx.fill()
  ctx.stroke()
}
