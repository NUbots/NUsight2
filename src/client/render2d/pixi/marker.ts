import { createTransformer } from 'mobx-utils'
import { Graphics } from 'pixi.js'

import { MarkerGeometry } from '../geometry/marker_geometry'
import { Shape } from '../object/shape'

import { applyAppearance } from './rendering'

export const renderMarker = createTransformer((shape: Shape<MarkerGeometry>): Graphics => {

  const { x, y, radius, heading } = shape.geometry

  const headingAngle = Math.atan2(heading.y, heading.x)
  const arcDistance = 3 * Math.PI * 0.5

  // By default, the arc startAngle begins on the positive x-axis and rotates clockwise. If the startAngle and
  // endAngle are offset by a quadrant, the arc will point toward the positive x-axis instead of starting there.
  const startAngleOffset = Math.PI * 0.25
  const startAngle = headingAngle + startAngleOffset
  const endAngle = headingAngle + arcDistance + startAngleOffset
  // The diagonal length of a unit square.
  const sqrt2 = Math.sqrt(2)

  const g = new Graphics()
  applyAppearance(g, shape.appearance, g => {
    g.arc(x, y, radius, startAngle, endAngle)
    g.lineTo(
      x + sqrt2 * radius * heading.x,
      y + sqrt2 * radius * heading.y,
    )
    g.closePath()
  })

  return g
})
