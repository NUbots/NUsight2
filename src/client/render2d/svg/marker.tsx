import { observer } from 'mobx-react'
import React from 'react'

import { Transform } from '../../math/transform'
import { MarkerGeometry } from '../geometry/marker_geometry'
import { Shape } from '../object/shape'

import { toSvgProps } from './rendering'

type Props = { model: Shape<MarkerGeometry>; world: Transform }
export const Marker = observer(
  ({
    model: {
      geometry: { x, y, radius, heading },
      appearance,
    },
  }: Props) => {
    const rotation = 135.0 + (180.0 / Math.PI) * Math.atan2(heading.y, heading.x)
    return (
      <path
        d="M-1 -1L0 -1A1 1 270 1 1 -1 0Z"
        transform={`scale(${radius}) rotate(${rotation}) translate(${x}, ${y})`}
        {...toSvgProps(appearance)}
      />
    )
  },
)
