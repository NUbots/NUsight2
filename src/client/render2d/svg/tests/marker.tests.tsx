import * as React from 'react'

import { Transform } from '../../../math/transform'
import { BasicAppearance } from '../../appearance/basic_appearance'
import { MarkerGeometry } from '../../geometry/marker_geometry'
import { Shape } from '../../object/shape'
import { Marker } from '../marker'

describe('MarkerSVGRenderer', () => {
  it('renders', () => {
    const model = Shape.of(MarkerGeometry.of(), BasicAppearance.of())
    const world = Transform.of()
    expect(<Marker model={model} world={world}/>).toMatchSnapshot()
  })
})
