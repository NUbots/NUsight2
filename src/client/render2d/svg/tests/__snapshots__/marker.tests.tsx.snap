// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkerSVGRenderer renders 1`] = `
<_class
  model={
    Shape {
      "appearance": BasicAppearance {
        "fillStyle": "#000",
        "lineWidth": 1,
        "strokeStyle": "#000",
      },
      "geometry": MarkerGeometry {
        "heading": Vector2 {
          "x": 1,
          "y": 0,
        },
        "radius": 1,
        "x": 0,
        "y": 0,
      },
      "group": Group {
        "children": Array [],
        "transform": Transform {
          "anticlockwise": true,
          "rotate": 0,
          "scale": Object {
            "x": 1,
            "y": 1,
          },
          "translate": Object {
            "x": 0,
            "y": 0,
          },
        },
      },
    }
  }
  world={
    Transform {
      "anticlockwise": true,
      "rotate": 0,
      "scale": Object {
        "x": 1,
        "y": 1,
      },
      "translate": Object {
        "x": 0,
        "y": 0,
      },
    }
  }
/>
`;
